internal/app/machined/pkg/controllers/files/iqn.go

```diff
@@ -0,0 +1,98 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+package files
+
+import (
+	"context"
+	"fmt"
+	"os"
+
+	"github.com/cosi-project/runtime/pkg/controller"
+	"github.com/cosi-project/runtime/pkg/safe"
+	"github.com/cosi-project/runtime/pkg/state"
+	"github.com/siderolabs/gen/optional"
+	"go.uber.org/zap"
+
+	runtimetalos "github.com/siderolabs/talos/internal/app/machined/pkg/runtime"
+	"github.com/siderolabs/talos/pkg/machinery/constants"
+	"github.com/siderolabs/talos/pkg/machinery/resources/files"
+)
+
+// IQNController creates an EtcFileSpec for the iSCSI Qualified Name (IQN) file.
+type IQNController struct {
+	V1Alpha1Mode runtimetalos.Mode
+}
+
+// Name implements controller.Controller interface.
+func (ctrl *IQNController) Name() string {
+	return "files.IQNController"
+}
+
+// Inputs implements controller.Controller interface.
+func (ctrl *IQNController) Inputs() []controller.Input {
+	return []controller.Input{
+		{
+			Namespace: files.NamespaceName,
+			Type:      files.EtcFileStatusType,
+			ID:        optional.Some("machine-id"),
+			Kind:      controller.InputWeak,
+		},
+	}
+}
+
+// Outputs implements controller.Controller interface.
+func (ctrl *IQNController) Outputs() []controller.Output {
+	return []controller.Output{
+		{
+			Type: files.EtcFileSpecType,
+			Kind: controller.OutputShared,
+		},
+	}
+}
+
+// Run implements controller.Controller interface.
+func (ctrl *IQNController) Run(ctx context.Context, r controller.Runtime, _ *zap.Logger) error {
+	// Skip the controller if we're running in a container.
+	if ctrl.V1Alpha1Mode == runtimetalos.ModeContainer {
+		return nil
+	}
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-r.EventCh():
+		}
+
+		// wait for machine-id to be available
+		_, err := safe.ReaderGetByID[*files.EtcFileStatus](ctx, r, "machine-id")
+		if err != nil {
+			if state.IsNotFoundError(err) {
+				continue
+			}
+
+			return fmt.Errorf("failed to get machine-id etcfile status: %w", err)
+		}
+
+		machineID, err := os.ReadFile("/etc/machine-id")
```

Comment:
I don't think it's correct to wait for a file and read it, we should instead use directly the Identity resource:

```go
		if err := safe.WriterModify(ctx, r, cluster.NewIdentity(cluster.NamespaceName, cluster.LocalIdentity), func(r *cluster.Identity) error {
			*r.TypedSpec() = localIdentity
```

https://github.com/siderolabs/talos/blob/399d53b543f6ca99f13d28313ae77b3472b0f728/internal/app/machined/pkg/controllers/cluster/node_identity.go#L103-L110

Reply:
fixed :+1: